akka-actor/src/main/scala/akka/pattern/internal/BackoffOnRestartSupervisor.scala

```diff
@@ -41,19 +41,19 @@ import scala.concurrent.duration._
           super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) => Escalate)
 
         strategy.decider.applyOrElse(ex, (_: Any) => defaultDirective) match {
-
           // Whatever the final Directive is, we will translate all Restarts
           // to our own Restarts, which involves stopping the child.
           case Restart =>
+            val nextRestartCount = restartCount + 1
+
             if (strategy.withinTimeRange.isFinite && restartCount == 0) {
               // If the user has defined a time range for the maxNrOfRetries, we'll schedule a message
               // to ourselves every time that range elapses, to reset the restart counter. We hide it
               // behind this conditional to avoid queuing the message unnecessarily
               val finiteWithinTimeRange = strategy.withinTimeRange.asInstanceOf[FiniteDuration]
-              system.scheduler.scheduleOnce(finiteWithinTimeRange, self, ResetRestartCount(restartCount))
+              system.scheduler.scheduleOnce(finiteWithinTimeRange, self, ResetRestartCount(nextRestartCount))
```

Comment:
The problem was that we'd send `restartCount` here which was still `0` and then increase the counter below just after triggering the restart, so when the `ResetRestartCount` arrived the counter would always be `1` and not get reset.